### ISSUE TITLE
Add metadata tokens/module versions back to reflection libraries in CORE CLR.

### ISSUE THREAD

**Issue by delmyers** [dotnet/corefx#2375]
Recent versions of the CORE clr have removed metadata from the reflection libraries. We are writing .NET diagnostics tools, and these APIs are important for supporting our diagnostics scenarios. We would like the APIs to be added back.

**Comment by nguerrera** [comment]
cc @AtsushiKan @davidwrighton @noahfalk @weshaggard @terrajobst 

Some issues that have come up before around these API:

1. They are not possible to implement on .NET Native

2. They have been abused (incorrect assumption that metadata token is stable across different versions and persisted as a key).

(2) is secondary -- we shouldn't exclude API just because they can be misused, particularly if there are no alternatives in .NET Core.

Questions I have:

1. How would we handle the .NET Native case: PlatformNotSupportedException? MissingMetadataException? Replace with TryGetMetadataToken(out int)?

2. Should we attack this at a higher level and consider other approaches for the MethodInfo -> Symbols diagnostic scenario (which from offline conversation is my understanding of what triggered this particular request.)

**Comment by noahfalk** [comment]
Adding one more question : )
Do you have the desire to support a similar scenario on .Net Native ever?

**Comment by rynowak** [comment]
:+1: for this on CoreCLR for the unit test support for DNX projects. We have some symbol resolution code that would be made much simpler with metadata token support.

> Should we attack this at a higher level and consider other approaches for the MethodInfo -> Symbols diagnostic scenario (which from offline conversation is my understanding of what triggered this particular request.)

In our case that's what we're really after. If something new is emerging from this, a unified story for desktop and CoreCLR would be great.

**Comment by noahfalk** [comment]
The current thinking is still relatively simple, adding some new extension methods in a new assembly that would provide the metadata tokens. For desktop it would be pretty easy to create the same extension methods that simply proxy to the pre-existing APIs if a particular tool wanted one set of API calls to target both stacks.

> Should we attack this at a higher level and consider other approaches for the MethodInfo -> Symbols
> diagnostic scenario (which from offline conversation is my understanding of what triggered this
> particular request.)

While a common high level abstraction would be convenient there is a fundamental challenge. The different runtime stacks are operating with different kinds and amounts of symbolic data. For example:
* Desktop/CoreCLR with an assembly+PDB loaded in-memory - The API could potentially expose PDB things like source line numbers
* Desktop/CoreCLR with assembly but no PDB - The API can expose things at MSIL level of abstraction such as tokens, but it would be impossible to expose source lines because the runtime doesn't have PDB data
* .Net Native - The API could only expose low level details like RVAs. Metadata tokens were lost during the compilation phase. (Although you could certainly envision a design that saves all the metadata tokens in the compiled binary, this is a performance tradeoff that .Net Native doesn't assume is always appropriate)

I'm open to better suggestions if anyone has thoughts. We may still do the small tactical solution for now but its always nice to have options.

**Comment by rynowak** [comment]
Our (ASP.NET) needs are just for Desktop/CoreCLR (no .Net Native).

The advantage that I see for something runtime that handles method -> symbol translation that it would handle all of the special cases correctly and centrally. The utility VS unit testing provides for this doesn't handle some special cases like async, requiring workarounds in xUnit's code. I just last night added support for it to our (ASP.NET/DNX) utility that does the same. 

For a running site we also provide source information on unhandled exceptions. I think that's just using information from the exception itself, which will include source/line info if the runtime has it.

**Comment by noahfalk** [comment]
When you are looking to translate to a 'symbol', what concretely do you have in mind? Transforming MethodInfo into Filename and line number?

**Comment by delmyers** [comment]
I agree with the direction of this conversation. What we are really looking for at the end of the day is source code navigation. Translation to file name and line number will not always be possible. What would be best is some API that, assuming we can access the correct pdb for the assembly, will give us the information that we need to read the code context from the pdb.

**Comment by noahfalk** [comment]
The current plan is to provide a metadata token (assuming the underlying runtime has one - CoreCLR does). It sounds like that satisfies your goal Del. I wasn't as clear if Ryan is shooting for something more.

**Comment by rynowak** [comment]
> The current plan is to provide a metadata token (assuming the underlying runtime has one - CoreCLR does). 

Adding the metadata token would unblock our current needs, and we'd still be happy if that's all you end up doing here.

Right now we're just looking up source/line information, but we're using MSDIA to do so, and have to have our own path to get the pdb for our loader and provide it to MSDIA. We're not super thrilled about owning that code, because in theory loading the symbol is duplicated with what the runtime already does behind the scenes. It would be an improvement if we could get access to a `Stream` or `byte[]` for the pdb provided by the runtime.

Also, we don't have a need right now for non-windows support for any of this functionality, but if anyone building xplat tools wanted parity on testing support we'd be without a story because of our MSDIA dependency.

**Comment by tmat** [comment]
@delmyers Could you shed some light on the scenario? in particular why do you start with Reflection objects in the first place? 

@rynowak By MSDIA, do you actually mean MSDIA or DiaSymReader (i.e. ISymUnmanagedReader interfaces). 

FYI: We are introducing new x-plat PDB format -- see https://github.com/tmat/roslyn/blob/master/docs/specs/PortablePdb-Metadata.md 

Please keep me in the loop so that your solutions work with the new format and don't need to be rewritten.

**Comment by rynowak** [comment]
We're using MSDIA (`IDiaSession` and friends) currently. We'd happily pivot on this though if it means we're aligning with everyone else. You can spy on us here: https://github.com/aspnet/Testing/blob/dev/src/Microsoft.Dnx.TestHost/

**Comment by tmat** [comment]
@rynowak Thanks. I'll take a look.

**Comment by delmyers** [comment]
Hey, Tomas, the immediate reason that we need this is that we need to put some code that is in ASP.NET MVC that will publish an event when MVC is about to call an action on a controller. Basically, it has to say "I'm going to call this method in this module". The message has to be lightweight, and it has to be published without a profiler or debugger attached, so metadata makes the most sense.

**Comment by tmat** [comment]
@delmyers Wouldn't sending a qualified member name be more robust?

**Comment by delmyers** [comment]
We thought about sending the method signature, but that ends up in a lot of compute time, reading a lot of reflected methods, types, etc, as well as doing a lot of string concatenation. Also, the consumer side is going to be expected to navigate to code. If we have the pdb information, and the assembly info, then we can easily use debuggers to do code navigation to the correct version of the code. If we have only the method name, then we have to parse the method name, and use a heuristic match for the version of the code that the user currently has.

**Comment by noahfalk** [comment]
@rynowak I think its reasonable to consider making the runtime's symbol data available when we have it (although I don't yet see anything that would make the work urgent/high priority). I would treat it as a separate feature with its own issue if you wanted to pursue it.

There is a big caveat that the runtime has never tried very hard to get symbol data and many tools would not be satisfied with such a limited search. The cases the runtime has symbol data are:
a) The PDB is located in the same directory as the binary (this doesn't always happen, for example windows UWP apps prohibit using the DIA APIs, so the runtime makes no load attempt)
b) The PDB was given to runtime directly as a byte array using Assembly.Load(byte[], byte[])
c) The PDB is being created on the fly in-memory by using reflection.emit.

The runtime never looks at hint paths, or symbol path environment variables, or any kind of symbol server. Its no problem if tools want to add those dependencies themselves but its not a dependency I'd want built-in. In general I'm trying to have the runtime depend as little as possible on symbols, and to hoist the dependencies out of the lower layers.

 If there are several tools that would all benefit from sharing code that understands how to find and manipulate symbols in common ways, you could also consider making a shared middle-ware library as another option?

**Comment by tmat** [comment]
@noahfalk I agree that the runtime should do minimal work. Perhaps just give access to the raw data it has or can easily find without implementing any lookup policies (or rather a trivial one -- next to the PE). 

As I pointed out here https://github.com/dotnet/corefx/issues/2768 we need access to metadata blob if available -- in cases a) and b) above. Similarly low level access to the PDB should be possible. Ideally as Stream or a memory pointer (ideally a read-only slice). 

The runtime indeed shouldn't load it if it doesn't need it. But when asked it should be able to provide it in the cases you mention. If the PDB is not collocated with the PE file (in-memory or on disk) then the caller should be responsible for finding it. It can do so using information in the PE debug directory. Such discovery should be implemented in a public library shared across tools that need it. The discovery might be specific to the system used by the customer, so there might be multiple implementations.

Once we move Reflection.Emit to Portable PDB format it can also provide the data as a blob of well defined format.

**Comment by delmyers** [comment]
Is there any movement on this ask? This thread has gone stale.

**Comment by delmyers** [comment]
Sorry. I accidentally hit the "close" button.

**Comment by nguerrera** [comment]
@delmyers There hasn't been much movement, but time is allocated to get the extension methods added for beta8. cc @weshaggard
